app/actions/handlers.py

```diff
@@ -111,82 +104,84 @@ async def action_pull_observations(integration, action_config: client.PullObserv
-                                    'action_id': "pull_observations"
-                                }
-                            )
-                            raise e
-                        else:
-                            if response:
-                                # Update states
-                                for vehicle in batch:
-                                    state = {
-                                        "latest_device_timestamp": vehicle.get("recorded_at"),
-                                        "mortality": vehicle["additional"].get("mortality")
+            if transformed_data:
+                # Send transformed data to Sensors API V2
+                def generate_batches(iterable, n=action_config.observations_per_request):
+                    for i in range(0, len(iterable), n):
+                        yield iterable[i: i + n]
+
+                for i, batch in enumerate(generate_batches(transformed_data)):
+                    async for attempt in stamina.retry_context(
+                            on=httpx.HTTPError,
+                            attempts=3,
+                            wait_initial=datetime.timedelta(seconds=10),
+                            wait_max=datetime.timedelta(seconds=10),
+                    ):
+                        with attempt:
+                            try:
+                                logger.info(
+                                    f'Sending observations batch #{i}: {len(batch)} observations. Device: {serial_num}'
+                                )
+                                await gundi_tools.send_observations_to_gundi(
+                                    observations=batch,
+                                    integration_id=integration.id
+                                )
+                            except httpx.HTTPError as e:
+                                msg = f'Sensors API returned error for integration_id: {str(integration.id)}. Exception: {e}'
+                                logger.exception(
+                                    msg,
+                                    extra={
+                                        'needs_attention': True,
+                                        'integration_id': str(integration.id),
+                                        'action_id': "pull_observations"
                                     }
-                                    await state_manager.set_state(
-                                        str(integration.id),
-                                        "pull_observations",
-                                        state,
-                                        vehicle.get("source"),
-                                    )
-                                    observations_extracted.append(response)
-        return {'observations_extracted': observations_extracted or []}
+                                )
+                                raise e
```

Comment:
Where does this exception get handled? From what I can tell, it looks like it will cause a retry at line 100 whereby the routine will essentially start from scratch. Does that seem true? And if it is, then it would mean a failure in posting to Gundi's sensors API will cause it to re-read from the ATS web service which will have advanced the data set cursor (meaning that we won't get the data that this logic is hoping to get).